src/compiler/srcloc.rs

```diff
@@ -29,19 +45,101 @@ pub struct Srcloc {
 
 impl Display for Srcloc {
     fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
-        match self.until {
+        match &self.until {
             None => formatter.write_str(&format!("{}({}):{}", self.file, self.line, self.col)),
-            Some((l, c)) => formatter.write_str(&format!(
+            Some(u) => formatter.write_str(&format!(
                 "{}({}):{}-{}({}):{}",
-                self.file, self.line, self.col, self.file, l, c
+                self.file, self.line, self.col, self.file, u.line, u.col
             )),
         }
     }
 }
 
 impl Srcloc {
+    pub fn new(name: Rc<String>, line: usize, col: usize) -> Self {
+        Srcloc {
+            file: name,
+            line,
+            col,
+            until: None,
+        }
+    }
+
+    pub fn ending(&self) -> Srcloc {
+        if let Some(u) = &self.until {
+            return Srcloc {
+                file: self.file.clone(),
+                line: u.line,
+                col: u.col,
+                until: None,
+            };
+        }
+        self.clone()
+    }
+
+    pub fn overlap(&self, other: &Srcloc) -> bool {
+        let mf: &String = self.file.borrow();
+        let of: &String = other.file.borrow();
+        if mf != of {
+            return false;
+        }
+
+        if self.line == other.line && self.col == other.col {
+            return true;
+        }
+
+        match (self.until.as_ref(), other.until.as_ref()) {
+            (None, None) => self.line == other.line && self.col == other.col,
+            (None, Some(_)) => other.overlap(self),
+            (Some(self_until), None) => {
+                if self.line < other.line && self_until.line > other.line {
+                    return true;
+                }
+                let len = self.len().unwrap_or(1);
+                if self.line == other.line && self.col <= other.col && self.col + len >= other.col {
```

Comment:
Should you be using `self_until` here instead of `self` ?
My reading is that `Some(x)` checks only matches when `self.until.as_ref()` is not `None`